casbin/model/model.cpp

```diff
@@ -0,0 +1,303 @@
+#include "model.h"
+#include "assertion.h"
+#include "../util/util.h"
+#include <iostream>
+using namespace std;
+
+map<string, string> sectionNameMap = {
+   {"r","request_definition"},
+   {"p","policy_definition"},
+   {"g","role_definition"},
+   {"e","policy_effect"},
+   {"m","matchers"}
+};
+
+const vector<string> requiredSections = { "r","p","e","m" };
+
+Model::Model()
+{
+	modelmap = {};
+}
+
+
+
+bool Model::AddDef(string sec, string key, string value)
+{
+	if (value == "")
+	{
+		return false;
+	}
+	Assertion ast = Assertion();
+	ast.Key = key;
+	ast.Value = value;
+
+
+	if (sec == "r" || sec == "p")
+	{
+
+		ast.Tokens = Util::Split(ast.Value, ",");
+		for (int i = 0; i< ast.Tokens.size(); i++)
+		{
+			ast.Tokens[i] = key + "_" + Util::Trim(ast.Tokens[i]," ");
+		}
+	}
+	else
+	{
+		ast.Value = Util::RemoveComments(Util::EscapeAssertion(ast.Value));
+	}
+
+	modelmap[sec][key] = ast;
+	return true;
+}
+
+
+
+void Model::BuildRoleLinks(RoleManager* rm)
+{
+	for (auto ast : modelmap["g"]) {
+		ast.second.buildRoleLinks(rm);
+	}
+}
+
+
+void Model::PrintPolicy()
+{
+
+}
+
+void Model::ClearPolicy() 
+{
+	for (auto& ast : modelmap["p"]) {
+		ast.second.Policy.clear();
+	}
+	for (auto& ast : modelmap["g"]) {
+		ast.second.Policy.clear();
+	}
+}
+
+vector<vector<string>> Model::GetPolicy(const string& sec, const string& ptype)
+{
+	return modelmap[sec][ptype].Policy;
+}
+
+bool Model::AddPolicy(const string& sec, const string& ptype, const vector<string>& rule)
+{
+	if (!HasPolicy(sec, ptype, rule))
+	{
+		modelmap[sec][ptype].Policy.push_back(rule);
+		return true;
+	}
+	return false;
+}
+
+bool Model::AddPolicies(const string& sec, const string& ptype, const vector<vector<string>>& rules)
+{
+	for (int i = 0; i < rules.size(); i++){
+		if (HasPolicy(sec, ptype, rules[i])) {
+			return false;
+		}
+	}
+
+	for (int i = 0; i < rules.size(); i++) {
+		modelmap[sec][ptype].Policy.push_back(rules[i]);
+	}
+	return true;
+}
+
+bool Model::HasPolicy(const string& sec, const  string& ptype, const vector<string>& rule)
+{
+	for (auto p : modelmap[sec][ptype].Policy)
+	{
+		if (Util::ArrayEquals(rule, p))
+			return true;
+	}
+	return false;
+}
+
+Model* Model::NewModelFromFile(const string& path) {
+	Model* m = new Model();
+	m->LoadModel(path);
+	return m;
+}
+
+Model*  Model::NewModelFromString(const string& text) {
+	Model* m =new Model();
+	m->LoadModelFromText(text);
+	return m;
+}

```

Comment:
I think these 2 methods are not necessary right? since you have the below 2 functions.